src/lib/prog_loaders.c

```diff
@@ -39,11 +40,167 @@ int prog_locate(struct prog *prog)
 	return 0;
 }
 
+/* Bits in HMER/HMEER */
+#define SPR_HMER_MALFUNCTION_ALERT	PPC_BIT(0)
+#define SPR_HMER_PROC_RECV_DONE		PPC_BIT(2)
+#define SPR_HMER_PROC_RECV_ERROR_MASKED	PPC_BIT(3)
+#define SPR_HMER_TFAC_ERROR		PPC_BIT(4)
+#define SPR_HMER_TFMR_PARITY_ERROR	PPC_BIT(5)
+#define SPR_HMER_XSCOM_FAIL		PPC_BIT(8)
+#define SPR_HMER_XSCOM_DONE		PPC_BIT(9)
+#define SPR_HMER_PROC_RECV_AGAIN	PPC_BIT(11)
+#define SPR_HMER_WARN_RISE		PPC_BIT(14)
+#define SPR_HMER_WARN_FALL		PPC_BIT(15)
+#define SPR_HMER_SCOM_FIR_HMI		PPC_BIT(16)
+#define SPR_HMER_TRIG_FIR_HMI		PPC_BIT(17)
+#define SPR_HMER_HYP_RESOURCE_ERR	PPC_BIT(20)
+#define SPR_HMER_XSCOM_STATUS		PPC_BITMASK(21,23)
+
+void read_scom_direct(uint32_t, uint64_t*);
+void read_scom_direct(uint32_t reg_address, uint64_t *buffer)
+{
+	asm volatile(
+		"ldcix %0, %1, %2":
+		"=r"(*buffer):
+		"b"(0x800603FC00000000),
+		"r"(reg_address << 3));
+	eieio();
+}
+
+void write_scom_direct(uint32_t, uint64_t*);
+void write_scom_direct(uint32_t reg_address, uint64_t *buffer)
+{
+	asm volatile(
+		"stdcix %0, %1, %2"::
+		"b"(*buffer),
+		"b"(0x800603FC00000000),
+		"r"(reg_address << 3));
+	eieio();
+}
+
+uint64_t read_hmer(void);
+uint64_t read_hmer(void)
+{
+	unsigned long val;
+	asm volatile("mfspr %0,%1" : "=r"(val) : "i"(0x150) : "memory");
+	return val;
+}
+
+#define PPC_BIT(bit)		(0x8000000000000000UL >> (bit))
+#define PPC_BITMASK(bs,be)	((PPC_BIT(bs) - PPC_BIT(be)) | PPC_BIT(bs))
+#define XSCOM_ADDR_IND_FLAG		PPC_BIT(0)
+#define XSCOM_ADDR_IND_ADDR		PPC_BITMASK(12,31)
+#define XSCOM_ADDR_IND_DATA		PPC_BITMASK(48,63)
+
+#define XSCOM_DATA_IND_READ		PPC_BIT(0)
+#define XSCOM_DATA_IND_COMPLETE		PPC_BIT(32)
+#define XSCOM_DATA_IND_ERR		PPC_BITMASK(33,35)
+#define XSCOM_DATA_IND_DATA		PPC_BITMASK(48,63)
+#define XSCOM_DATA_IND_FORM1_DATA	PPC_BITMASK(12,63)
+#define XSCOM_IND_MAX_RETRIES 10
+
+void write_scom_indirect(uint64_t, uint64_t);
+void write_scom_indirect(uint64_t reg_address, uint64_t value)
+{
+	uint64_t addr;
+	uint64_t data;
+	addr = reg_address & 0x7FFFFFFF;
+	data = reg_address & XSCOM_ADDR_IND_ADDR;
+	data |= value & XSCOM_ADDR_IND_DATA;
+	printk(BIOS_EMERG, "----------------\n");
+	printk(BIOS_EMERG, "SCOM WRITE\n\n");
+	printk(BIOS_EMERG, "hmer = %llX, fail = %llX, done = %llX, status = %llX\n",
+		read_hmer(),
+		read_hmer() & SPR_HMER_XSCOM_FAIL,
+		read_hmer() & SPR_HMER_XSCOM_DONE,
+		read_hmer() & SPR_HMER_XSCOM_STATUS);
+	printk(BIOS_EMERG, "WRITING addr = %llX, data = %llX\n", addr, data);
+	write_scom_direct(addr, &data);
+	printk(BIOS_EMERG, "hmer = %llX, fail = %llX, done = %llX, status = %llX\n",
```

Comment:
Maybe try to repeat whole `printk` a couple of times, this may be a timing issue: https://github.com/3mdeb/talos-hostboot/blob/master/src/usr/scom/scom.C#L489